aten/src/ATen/native/LossCTC.cpp

```diff
@@ -0,0 +1,313 @@
+  auto target_lengths_arg = TensorArg(target_lengths, "target_lengths", 4);
+  checkScalarType(c, targets_arg, kLong);
+  checkScalarType(c, input_lengths_arg, kLong);
+  checkScalarType(c, target_lengths_arg, kLong);
+  checkDim(c, log_probs_arg, 3);
+  checkDimRange(c, targets_arg, 1, 3);
+
+  int64_t batch_size = log_probs.size(1);
+  int64_t num_labels = log_probs.size(2);
+  AT_CHECK(BLANK < num_labels, "blank must be in label range");
+  checkSize(c, input_lengths_arg, {batch_size});
+  checkSize(c, target_lengths_arg, {batch_size});
+
+  size_t tg_target_stride;
+  int64_t max_target_length;
+  Tensor tg_batch_offsets;
+  if (targets.dim() == 1) { // concatenated targets
+    tg_batch_offsets = at::zeros_like(target_lengths);
+    auto tmp = tg_batch_offsets.narrow(0, 1, batch_size-1);
+    at::cumsum_out(tmp, target_lengths, 0);
+    max_target_length = at::max(target_lengths).toCLong();
+    tg_target_stride = targets.stride(0);
+    checkSize(c, targets_arg, 0, tmp[batch_size-1].toCLong()+target_lengths[batch_size-1].toCLong());
+  }
+  else { // batch x max_target_length
+    // dim is 2
+    tg_batch_offsets = at::arange(0, targets.stride(0)*batch_size, targets.stride(0), target_lengths.options());
+    tg_target_stride = targets.stride(1);
+    max_target_length = targets.size(1);
+    checkSize(c, targets_arg, 0, batch_size);
+    AT_CHECK(targets.size(1) >= max_target_length,
+             "Expected tensor to have size at least ", max_target_length, " at dimension 1, but got size ", targets.size(1), " for ", targets_arg,
+             " (while checking arguments for ", c, ")");
+  }
+  int max_input_length = at::max(input_lengths).toCLong();
+  AT_CHECK(log_probs.size(0) >= max_input_length,
+           "Expected tensor to have size at least ", max_input_length, " at dimension 1, but got size ", targets.size(0), " for ", targets_arg,
+           " (while checking arguments for ", c, ")");
+
+  Tensor log_alpha = at::empty({batch_size, log_probs.size(0), 2*max_target_length+1}, log_probs.options());
+  Tensor neg_log_likelihood = at::empty({batch_size}, log_probs.options());
+
+
+  auto log_probs_data = log_probs.data<scalar_t>();
+  auto log_alpha_data = log_alpha.data<scalar_t>();
+  auto targets_data = targets.data<int64_t>();
+  auto neg_log_likelihood_data = neg_log_likelihood.data<scalar_t>(); // we assume stride one for the only dimension for freshly allocated tensor
+  size_t lp_input_stride = log_probs.stride(0);
+  size_t lp_char_stride = log_probs.stride(2);
+  size_t la_input_stride = log_alpha.stride(1);
+  size_t la_target_stride = log_alpha.stride(2);
+
+  log_alpha.narrow(1, 0, 1).fill_(neginf); // or do this inside the batch loop?
+  #pragma omp parallel for
+  for (int64_t b = 0; b < batch_size; b++) {
+    int64_t input_length = input_lengths[b].toCLong();
+    int64_t target_length = target_lengths[b].toCLong();
+    int64_t lp_batch_offset = b*log_probs.stride(1);
```

Comment:
you can save the strides before the loop too. :)

Reply:
Will do. I never know which bits are expensive and which aren't...
Thank you for looking at this!

Reply:
`.stride()` is kinda expensive on master (see #9732 ), but will be fast eventually. However, it would be good to still cache them.

`arr[i].toCLong()` on the other hand, is always much slower than `arr_data[i]`.